Use the chained router to find root pages

aschempp opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2499

I had this issue in a client application, debugging it showed that the fallback 404 was actually found correctly, but `Frontend::getRootPageFromUrl` used by  `PageError404` did not resolve correctly. Most likely because the error404 pages are not present in the regular matcher but only in the chained router/matcher.

aschempp committed:
Use router to find root page

aschempp commented:
The failing tests look like an issue in the GitHub Actions?

m-vo commented:
> The failing tests look like an issue in the GitHub Actions?

Yes, looks similar in #2716. Except for the 7.4 + 8.0 test.

leofeyer committed:
Merge branch '4.9' into bugfix/fallback-404

leofeyer commented:
I have fixed the unrelated failures. The remaining failures are related to the PR, I'm afraid. 🙈

aschempp committed:
Correctly match 404 page and unknown languages

aschempp commented:
I have updated the implementation. The failing test showed why using the router is not ideal:
 1. it will use _all_ routers, e.g. also local and custom ones. That means it _might_ match a root page that does not come from Contao. While that is correct, it is not what the method is looking for, so using the matcher is better.
 2. **because** it uses the router, it also runs the route enhancers, which modify the `$_GET` parameters based on matched results. We certainly don't want that here though.

I also added a test for the current issue. This lead me to notice we only check a 404 error is thrown (the Symfony one) and not actually rendering the Contao 404 case. This is now all resolved.

aschempp committed:
Fixed config test

aschempp committed:
CS

leofeyer commented on core-bundle/src/Resources/contao/classes/Frontend.php:
```diff
@@ -332,24 +332,32 @@ public static function getRootPageFromUrl()
+		$objRootPage = $arrParameters['pageModel'] ?? null;
 
-			throw new NoRootPageFoundException('No root page found', 0, $exception);
+		if (!$objRootPage instanceof PageModel)
```
Why not `if (!isset($arrParameters['pageModel'])) {`? Then we would not need the `?? null` part. And the error message would be more accurate, because now `$objRootPage` could be a non-PageModel object and the error would still be "must have a pageModel parameter".

aschempp replied:
I think mostly because that tells my IDE that `$objRootPage` is a `PageModel` later in the code 🤔

aschempp replied:
also, this has been the code before, that didn't change (only rearranged) 😉
